### ISSUE TITLE
Concatenating tensors in network architecture causes slow down in training speed

### ISSUE THREAD

**Issue by barisumog** [keras-team/keras#13084]
I understand there'll be some overhead when concatenating, but I think the current slow down is beyond what should be expected.

I've been experiencing this with various network setups, both on my own machine and on google compute servers.

Below is a small script to display the difference in training time. Both networks are essentially the same, with same number of parameters. But using concatenate introduces 50% slow down in this example. I experience similar or greater rate of slow down in my actual work when using concatenation (compared to same parameter count with no concatenation).

**System information**  
- OS Platform and Distribution: Ubuntu 18.04
- TensorFlow backend:  Yes
- TensorFlow version:  1.12.0 & 1.13.1
- Keras version:  2.2.4
- Python version:  3.6.8 & 3.7.3
- CUDA/cuDNN version:  9.0 / 7.3.1
- GPU model and memory:  GTX 850M 2gb & Tesla P100 16gb

Simple network without concatenation:
```
import numpy as np
import keras

X = np.random.random((2048, 128, 128, 1)).astype('float32')
Y = (np.random.random(2048) < 0.5).astype('int8')

inp = keras.layers.Input(shape=(128, 128, 1))
net = keras.layers.Conv2D(
    128, (3, 3), padding='same', strides=(1, 1), activation='relu')(inp)
net = keras.layers.Flatten()(net)
net = keras.layers.Dense(1, activation='sigmoid')(net) 

model = keras.models.Model(inp, net)
model.compile(keras.optimizers.Adadelta(), loss='binary_crossentropy')
model.summary()

model.fit(X, Y, epochs=5, shuffle=True, verbose=2, batch_size=8)
```

Output:
```
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         (None, 128, 128, 1)       0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 128, 128, 128)     1280      
_________________________________________________________________
flatten_1 (Flatten)          (None, 2097152)           0         
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 2097153   
=================================================================
Total params: 2,098,433
Trainable params: 2,098,433
Non-trainable params: 0
Epoch 1/5
 - 19s - loss: 7.7482
Epoch 2/5
 - 17s - loss: 7.7688
Epoch 3/5
 - 18s - loss: 7.7688
Epoch 4/5
 - 18s - loss: 7.7688
Epoch 5/5
 - 17s - loss: 7.7688
```

Network changed to using concatenation (rest of code is same as above):
```
inp = keras.layers.Input(shape=(128, 128, 1))
combo = []
for i in range(16):
    net = keras.layers.Conv2D(
        8, (3, 3), padding='same', strides=(1, 1), activation='relu')(inp)
    combo.append(net)
net = keras.layers.Concatenate()(combo)
net = keras.layers.Flatten()(net)
net = keras.layers.Dense(1, activation='sigmoid')(net)
```

Output:
```
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_1 (InputLayer)            (None, 128, 128, 1)  0                                            
__________________________________________________________________________________________________
conv2d_1 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_2 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_3 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_4 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_5 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_6 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_7 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_8 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_9 (Conv2D)               (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_10 (Conv2D)              (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_11 (Conv2D)              (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_12 (Conv2D)              (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_13 (Conv2D)              (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_14 (Conv2D)              (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_15 (Conv2D)              (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
conv2d_16 (Conv2D)              (None, 128, 128, 8)  80          input_1[0][0]                    
__________________________________________________________________________________________________
concatenate_1 (Concatenate)     (None, 128, 128, 128 0           conv2d_1[0][0]                   
                                                                 conv2d_2[0][0]                   
                                                                 conv2d_3[0][0]                   
                                                                 conv2d_4[0][0]                   
                                                                 conv2d_5[0][0]                   
                                                                 conv2d_6[0][0]                   
                                                                 conv2d_7[0][0]                   
                                                                 conv2d_8[0][0]                   
                                                                 conv2d_9[0][0]                   
                                                                 conv2d_10[0][0]                  
                                                                 conv2d_11[0][0]                  
                                                                 conv2d_12[0][0]                  
                                                                 conv2d_13[0][0]                  
                                                                 conv2d_14[0][0]                  
                                                                 conv2d_15[0][0]                  
                                                                 conv2d_16[0][0]                  
__________________________________________________________________________________________________
flatten_1 (Flatten)             (None, 2097152)      0           concatenate_1[0][0]              
__________________________________________________________________________________________________
dense_1 (Dense)                 (None, 1)            2097153     flatten_1[0][0]                  
==================================================================================================
Total params: 2,098,433
Trainable params: 2,098,433
Non-trainable params: 0
Epoch 1/5
 - 28s - loss: 8.0518
Epoch 2/5
 - 26s - loss: 8.0802
Epoch 3/5
 - 27s - loss: 8.0802
Epoch 4/5
 - 26s - loss: 8.0802
Epoch 5/5
 - 28s - loss: 8.0802
```

**Comment by jvishnuvardhan** [comment]
@barisumog I think the reason behind higher computation time (even-though the number of trainable params are same) is higher number of layers which takes more time for backprop algorithm (several gradient estimations, etc). I reduced the number of concatenated layers by half and increase number of params by two so that the trainable params are same, then the computational time is comparable to non-concatenated model. Please check the [gist here]( https://colab.sandbox.google.com/gist/jvishnuvardhan/840e742252ccc0f5f3e2528ea6e99e28/keras_13084.ipynb). thanks!

**Comment by barisumog** [comment]
@jvishnuvardhan Thanks for the info. How about a setup like this?

```
inp = keras.layers.Input(shape=(128, 128, 1))
net = keras.layers.Conv2D(
    16, (3, 3), padding='same', strides=(1, 1), activation='relu')(inp)
for i in range(7):
    new = keras.layers.Conv2D(
        16, (3, 3), padding='same', strides=(1, 1), activation='relu')(inp)
    net = keras.layers.Concatenate()([net, new])
net = keras.layers.Flatten()(net)
net = keras.layers.Dense(1, activation='sigmoid')(net)
```

This has 8 separate conv layers, same as in your gist, but with more concatenations. Same # of params as before. An epoch takes around 28 secs on my computer, again 50% slower compared to original one. So it seems to me, with same number of conv layers, more concats mean significantly more slow down.

Shouldn't # of backprop paths be dependent on # of conv layers? Shouldn't 8 conv layers and 1 concat take _almost_ the same time as 8 conv layers and 7 concat?

Sorry if I'm mistaken, or if this doesn't make sense. I appreciate your time and help. Thanks.
